Add OceanBC interface modeled on AtmosBC interface

blallen opened the pull request:
<!--
Thanks for submitting code to CLIMA, the Climate Machine.

Before continuing, please be sure you have:

1. Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
2. Followed all necessary [style guidelines](https://climate-machine.github.io/CLIMA/latest/CodingConventions.html)
3. Identified key contributors to review this submission
-->

# Description

Updating the Ocean boundary conditions interface to match that used in Atmos. Makes specifying boundary conditions easier for users and should make the move to a cubed sphere easier. 

<!--- Please leave the following section --->

# For review by CLIMA developers

- [x] There are no open pull requests for this already
- [x] CLIMA developers with relevant expertise have been assigned to review this submission
- [x] The code conforms to the [style guidelines](https://climate-machine.github.io/CLIMA/latest/CodingConventions.html) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [x] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

blallen commented:
@simonbyrne here's my first pass at matching ocean boundary conditions to your interface. Everything runs right now, but the boundary conditions aren't being applied correctly yet. I'll be working on that throughout the week

@jkozdon and @sandreza, can you check the `bc_temperature.jl` and `bc_velocity.jl` files to make sure I'm actually applying everything correctly and in correspondence with the design docs?

blallen commented:
Just rebased on top of master and got this fresh again.

Three big improvements beyond that:

1) implement Neumann boundary conditions only using boundary states. @simonbyrne take a look at this and let me know if it will also work for the atmosphere. Having one place to implement all BCs should be easier for developers.

2) Realized that the ocean surface can't be `Impenetrable` in a hydrostatic boussinesq set up, so I implemented a `Penetrable` mass/velocity boundary condition with free slip and kinematic stress options.

3) Started working to find a nice interface for specifying different physical boundary fluxes on different boundaries. This part is still WIP and can be factored out into a separate PR.

blallen commented:
bors try

blallen commented:
@jm-c @christophernhill, here's the new interface for specifying boundary conditions for the ocean. Things are split out into velocity and temperature boundary conditions, so you can mix and match between the two. Also, the boundary conditions should apply in any geometry, not just the box (assuming the `kinematic_stress` and `surface_flux` functions are general enough). 

@simonbyrne for this round, I'm not letting users pass functions to the flux boundary conditions, so each boundary will get the same flux forcings. I have some WIP in progress on passing a function like you do with AtmosBC, but that will come in a separate PR. I want to get this stuff in because it's been working for a while.

also @simonbyrne, take a look at how I did the flux boundary conditions and let me know if you think this makes sense for the Atmos model as well. @charleskawczynski and @gagnelandmanna have already implemented similar approaches in the soil models.

blallen commented:
bors try

blallen commented:
bors try-

blallen commented:
bors try

charleskawczynski commented on src/Ocean/HydrostaticBoussinesq/BoundaryConditions.jl:
```diff
@@ -3,6 +3,20 @@ using ..DGmethods.NumericalFluxes:
+    velocity::M = Impenetrable(NoSlip())
+    temperature::T = Insulating()
+end
+
```
```suggestion
```
We can’t have new lines between doc strings and structs/functions

charleskawczynski commented on src/Ocean/HydrostaticBoussinesq/BoundaryConditions.jl:
```diff
@@ -3,6 +3,20 @@ using ..DGmethods.NumericalFluxes:
+
+The standard boundary condition for OceanModel. The default options imply a "no flux" boundary condition.
+"""
+
```
```suggestion
```
We can’t have new lines between doc strings and structs/functions

charleskawczynski commented on src/Ocean/HydrostaticBoussinesq/bc_velocity.jl:
```diff
@@ -0,0 +1,330 @@
+apply no slip boundary condition for velocity
+sets reflective ghost point
+"""
+
```
```suggestion
```

jkozdon commented:
> No opinion on this one.

But seems like a good idea to have a unified interface. Thanks!

blallen commented:
Bors r+

blallen commented:
Just to note, this PR fixes a mistake introduced in PR #660. When that PR switched us from a central boundary flux formulation 

https://github.com/CliMA/ClimateMachine.jl/blob/461657afa53ad608015a27f7c91dd8c15e605810/src/DGmethods/NumericalFluxes.jl#L244-L268

to an inflow boundary flux formulation

https://github.com/CliMA/ClimateMachine.jl/blob/d808788447a2354d1e5927b986d212780f078278/src/DGmethods/NumericalFluxes.jl#L241-L263

for the boundary flux for the second-order/diffusive fluxes, we should have changed 

https://github.com/CliMA/ClimateMachine.jl/blob/c9bc1b6598d1340b238f5c22b85ae745f8954ee5/src/Ocean/Model/HydrostaticBoussinesqModel.jl#L463
and
https://github.com/CliMA/ClimateMachine.jl/blob/c9bc1b6598d1340b238f5c22b85ae745f8954ee5/src/Ocean/Model/HydrostaticBoussinesqModel.jl#L470

to `D⁺.∇u = Diagonal(A⁺.ν) \  τ` and `D⁺.∇θ = Diagonal(A⁺.κ) \  σ` to match. (The left divides got reverted in a PR #841 when we moved the viscosity and diffusivities back into the gradient fluxes). 

This omission caused the code to apply double the prescribed wind stress and temperature forcing during the period between PR #600 (Feb 5 2020) and PR #851 (May 30 2020). When updating the boundary condition treatment for this PR, I corrected the mistake.
